File: solved/add-binary.ts
/**
 * https://leetcode.com/problems/add-binary/
 */
export { addBinary };

function addBinary(a: string, b: string): string {
  let carry = 0;
  let longestLength = Math.max(a.length, b.length);
  let solution: string[] = Array(longestLength);
  let aLastIndex = a.length - 1;
  let bLastIndex = b.length - 1;

  for (let i = 0; i < longestLength; i++) {
    let aInt = toInt(a[aLastIndex - i]);
    let bInt = toInt(b[bLastIndex - i]);
    let sum = aInt + bInt + carry;

    if (sum > 1) {
      if (sum === 2) solution[i] = '0';
      else solution[i] = '1';
      carry = 1;
    } else {
      if (sum === 1) solution[i] = '1';
      else solution[i] = '0';
      carry = 0;
    }
  }

  if (carry > 0) solution.push('1');

  return solution.reverse().join('');
}

function toInt(x: string | undefined): number {
  if (x === undefined || x === '0') return 0;
  else return 1;
}

File: returned/python/today5/add_binary.py
# https://leetcode.com/problems/add-binary/

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        index_a = len(a) - 1
        index_b = len(b) - 1
        ans = []
        carry = 0
        while index_a >= 0 or index_b >= 0 or carry == 1:
            digit_a, digit_b = 0, 0
            if index_a >= 0:
                digit_a = int(a[index_a])
            if index_b >= 0:
                digit_b = int(b[index_b])

            total = digit_a + digit_b + carry
            if total <= 1:
                ans.append(str(total))
                carry = 0
            else:
                ans.append(str(total % 2))
                carry = 1

            index_a -= 1
            index_b -= 1
        return ''.join(reversed(ans))

        